Fix Product version string invalidator triggers translation warnings

RadoslavGeorgiev opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-6053.

(For Bug Fixes) Bug introduced in PR https://github.com/woocommerce/woocommerce/pull/61613.

- Use `RegisterHooksInterface` and call `->register()` from the main plugin class for consistency.
- Instead of trying to immediately attach all hooks, do it on `init`. I verified that all other hooks run later, so this should be a functional approach.
- Fix the unit test. Since the change is minimal, the only change there is to call `->register()` when needed.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Follow the testing instructions from https://github.com/woocommerce/woocommerce/pull/61613, as this PR does not change any behavior.

<!-- End testing instructions -->

### Testing that has already taken place:

The change is relatively simple, and I don't have the proper caching plugins to test it locally. I relied on unit tests to confirm that everything works as expected.

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

This PR addresses an issue from another PR that was only merged a day before this one got created.

</details>

RadoslavGeorgiev committed:
Add the hooks interface, attach everything on `init`.

RadoslavGeorgiev committed:
Fix tests

RadoslavGeorgiev committed:
Fix typo

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -54,7 +65,17 @@ final public function init( FeaturesController $features_controller ): void {
 	 * @return void
 	 */
-	private function register_hooks(): void {
+	public function on_init(): void {
```
This is a change in a freshly introduced PR, there is no backwards compatibility to break.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -13,13 +14,20 @@
  * any cached REST API responses containing that product.
  */
-class ProductVersionStringInvalidator {
+class ProductVersionStringInvalidator implements RegisterHooksInterface {
```
This is a change in a freshly introduced PR, there is no backwards compatibility to break.

FYI, `RegisterHooksInterface` is an internal interface, so the comment about the class following public interface contracts is moot.

RadoslavGeorgiev committed:
Merge branch 'trunk' into wooplug-6053-product-version-string-invalidator-triggers-translation

senadir commented:
Came to report this, thank you for working on it @RadoslavGeorgiev

senadir committed:
Remove stale PHPStan baseline entry

senadir commented:
Fixed the phpstan issue and confirmed this is fixing the issue.

senadir committed:
Merge branch 'trunk' into wooplug-6053-product-version-string-invalidator-triggers-translation

Konamiman reviewed (changes requested):
Many thanks for providing a fix while I'm afk.

However, I'd prefer a simpler approach that doesn't require the hooks to be registered within the `init` hook: what if we just check if the `woocommerce_feature_rest_api_caching_enabled` option has the value `yes` as a direct one-line replacement for the `$features_controller->feature_is_enabled( 'rest_api_caching' )` check? We definitely need to go and take a look at `FeaturesController` and make it support this kind of "check feature in class initializer" scenarios, but meanwhile, let's keep things simple please.

Also there's no need to implement `RegisterHooksInterface`: this was used by the old DI container but the new one doesn't use it, so that's another opportunity to keep things simple.

mcliwanow commented:
@RadoslavGeorgiev, while working on another PR, I noticed that to "get rid" of the textdomain warning, I need to update both `ProductVersionStringInvalidator` AND `CartEventTracker ` (merged just 2 days ago in https://github.com/woocommerce/woocommerce/pull/62543, cc @leandroalonso). It follows a similar pattern of checking features too early.

Do you think we can update`CartEventTracker` too in your PR? I wanted to put together a quick PR but based on the comments I see that there are some concerns about using the `init` hook.

> what if we just check if the woocommerce_feature_rest_api_caching_enabled option has the value yes as a direct one-line replacement for the $features_controller->feature_is_enabled( 'rest_api_caching' ) check?

@Konamiman, this assumes that `enabled_by_default` (part of feature flag definition) is `false`, right? For this particular feature flag it's a valid assumption but if it ever changes, we might forget about this option check 🤔 Just my two cents

Konamiman commented:
@mcliwanow 

> this assumes that enabled_by_default (part of feature flag definition) is false, right? For this particular feature flag it's a valid assumption but if it ever changes, we might forget about this option check

Good point, but that's ok because this is a temporary fix (and the feature is experimental anyway, so ongoing changes are fine). I'll research how a proper fix could be implemented for `FeaturesController` itself.

Konamiman commented:
https://github.com/woocommerce/woocommerce/pull/62665 implements the quick fix of checking the feature option directly instead of using `FeaturesController::feature_is_enabled` as a temporary fix. We'll later decide if we merge this pull request as a more permanent fix, or https://github.com/woocommerce/woocommerce/pull/62662, or something else.

RadoslavGeorgiev committed:
Drop the RegistersHooksInterface

RadoslavGeorgiev committed:
Use options directly, rather than the features controller

RadoslavGeorgiev commented:
I should have checked my inbox before addressing your comment 🤷 

This PR is now also updated to check for the option directly in a more simplified manner. The only remaining difference is is the call to `->register()` rather than adding hooks within the constructor. You are welcome to pick either one to merge and close the other @Konamiman.

I'm also personally ok with your approach to provide callbacks in #62662 (cool PR number, by the way). That would let features hook in earlier than `init` when required.

Konamiman commented:
@RadoslavGeorgiev Let's do this to avoid (or at least minimize) confussion:

1. I'll merge https://github.com/woocommerce/woocommerce/pull/62665.
2. Please undo your latest update to this pr (so that it goes back to implementing your original idea of deferring the initialization) to after `init`, and turn it back into draft.
3. Let's turn https://github.com/woocommerce/woocommerce/pull/62662 into draft too.

Later on we'll discuss which approach to follow, merge the "winning" pr, and close the other one (or maybe we'll do something entirely different and close both).

RadoslavGeorgiev committed:
Merge branch 'trunk' into wooplug-6053-product-version-string-invalidator-triggers-translation

RadoslavGeorgiev committed:
Revert "Use options directly, rather than the features controller"

This reverts commit 8b727d8665166e67a50e128147139ed7fe81a69e.
